Initial SMP implementation

dancrossnyc opened the pull request:
Provide an initial SMP implementation.  This is hard-coded to start up 4 CPUs, as in the test configuration, but is enough to validate that the SMP startup logic works.

Signed-off-by: Dan Cross <cross@gajendra.net>

dancrossnyc commented:
Hey Mike, Cory...Could I please trouble either or both of you for a peek at this code?

waych commented:
Looks pretty good Dan!

Not going to lie, I got a little confused by the math at the start of setup_sipi_page() which is effectively the math behind APSTART in the assembler. Better that this were separated.

I wonder about the uses of leak().  For the stacks it sorta makes sense, but it feels reasonable that maybe the others could or should be recovered?  I don't know. Hard to reason given that theon is expected to be ephemeral itself, but then the question still stands, are the stacks expected to be discoverable across cores somewhere (for debug), and if so is that tied to these initial AP bootstrap resources that themselves are ephemeral?  I guess I'm assuming a broadcast SIPI would not be used for subsequent upboots, but I'm not sure that matters to the question.

I tried running this quickly locally to see how that'd work.  I haven't booted this in a while.  Currently for me this seems to get hung up after printing `regions:` before this commit and gets further along but dies printing out `ELF for "devices" $(Segment@HPA(` which suggests something funny is going on?  I haven't dug in.

dancrossnyc commented:
> Looks pretty good Dan!

Thanks, Mike!

> Not going to lie, I got a little confused by the math at the start of setup_sipi_page() which is effectively the math behind APSTART in the assembler. Better that this were separated.

Oh sure, I can change that around and try to clarify it. Would it help to make all the arithmetic in terms of byte pointers, or would it help to extract that to a function all its own, or something else entirely? I've done the former for now, and also added a lot of explanatory comments. Let me know what you'd like, though.

> I wonder about the uses of leak(). For the stacks it sorta makes sense, but it feels reasonable that maybe the others could or should be recovered? I don't know. Hard to reason given that theon is expected to be ephemeral itself, but then the question still stands, are the stacks expected to be discoverable across cores somewhere (for debug), and if so is that tied to these initial AP bootstrap resources that themselves are ephemeral? I guess I'm assuming a broadcast SIPI would not be used for subsequent upboots, but I'm not sure that matters to the question.

So I think there are two separable issues here.

First, this code specifically, which is hard-coding 4 CPUs and leaking their stacks just for the purposes of testing and demonstrating that we can do something in the immediate term, but that's obviously not reasonable and will have to be fixed. For the time being I've added some comments to `fn cpus()` in `main.rs` saying as much.

The second issue is essentially what you've raised: to what extent at stacks discoverable from other CPUs? For isolation reasons, ideally "not at all", but for effective debugging, we've got to have something; I confess I haven't given too much thought to this. It occurs to me that leaking the stacks (and the CPU list) is somewhat orthogonal: that's just Rust machinery to keep the Box and Vec drop code from reclaiming the memory, but we're retaining pointers to both: the `EntryCPU` array we've dropped a pointer to in the SIPI page retains a pointer to the Vec, and each entry retains a pointer to the (top) of the stack, so we could get access to all of that data again.

We'll have to retain a list of all CPUs in the system in the global segment eventually, and we could certainly store the physical address of each CPU's scheduler stack in there, which seems like a reasonable way to support poking at that memory when we do something for debugging.

> I tried running this quickly locally to see how that'd work. I haven't booted this in a while. Currently for me this seems to get hung up after printing `regions:` before this commit and gets further along but dies printing out `ELF for "devices" $(Segment@HPA(` which suggests something funny is going on? I haven't dug in.

Huh, weird. I'm guessing you're starting it via, `cargo xtask qemukvm` ? I haven't seen that, but I'm sure it's something I'm messing up.

Oh that reminds me; that's another thing: once we submit this we can't run it with QEMU except under KVM, since it needs the x2APIC.

waych commented:
> Would it help to make all the arithmetic in terms of byte pointers, or would it help to extract that to a function all its own, or something else entirely?

I'm sorta ambivalent. I liked the pointer arith with the usize * but I did look at it and have a quick moment to wonder what was happening.  Frankly I think it is clearly than the 8 and 16 byte offsets.

Re: the page 7, I guess seeing the math right there mixed in with the HPA page slice construction got me mixed up.  The new comments add a lot to clear things up. Reading the original code I was left wondering "why page 7" and I guess that is somewhat arbitrary, though there aren't that many options I guess.

> via, `cargo xtask qemukvm` ?

Yeah. I was doing cargo xtask qemu at first and realized quickly that didn't work.

I did a little bit of debug to see what was up.  Looks maybe like a trap storm or some such?

waych commented:
fwiw I think the stack got smashed, sadly due to using the default debug build.

Running info registers repeatedly while hung I see the RIP fluctuate throughout `trap()`, but **also** I see it sometimes as 0000000000000008 which is clearly not correct. Not sure where this is coming from but it explains what the cpu is doing.  Looking at the stack however, it looks like it isn't smashed, the RIP looks like its in a good spot relative to the bootstack page, however:

```
(qemu) x/512xg 0xffff8000001a4000
ffff8000001a4000: 0xffff8000001a7650 0x000000000000000e
theon heap^
ffff8000001a4010: 0x0000000000000000 0x0000000000000000
ffff8000001a4020: 0x00ff8000001a41f0 0xffff800000171ed5
------------------------------------------^
_ZN4core3ptr9const_ptr33_$LT$impl$u20$$BP$const$u20$T$GT$13is_aligned_to17hdde12cb5b1a00deaE:
calling:
_ZN4core3ptr9const_ptr33_$LT$impl$u20$$BP$const$u20$T$GT$13is_aligned_to12runtime_impl17h80808c6ee66b3582E

ffff8000001a4030: 0xffff8000001a4160 0x0000000000000008
bootstack
ffff8000001a4040: 0xffff8000001a41f0 0xffff8000001a4c78
bootstack bootstack
ffff8000001a4050: 0x0000000000000001 0xffff800000198fae
----------------------------------------UNKNOWN^^   Above HEAP?

ffff8000001a4060: 0x0000000000000000 0xffff800000198fae
ffff8000001a4070: 0x0000000000000001 0xffff800000198fae
ffff8000001a4080: 0xffff800000198fae 0xffff800000171ba5
_ZN4core3ptr9const_ptr33_$LT$impl$u20$$BP$const$u20$T$GT$13is_aligned_to17h186dbae12c942f3eE

ffff8000001a4090: 0x0000000000000008 0x0000000000000001
ffff8000001a40a0: 0xffff800000198fae 0x0000000000000008
ffff8000001a40b0: 0x0000000000000001 0xffff800000171a8f
_ZN4core3ptr9const_ptr33_$LT$impl$u20$$BP$const$u20$T$GT$10is_aligned17h3456a7a0f7787080E

ffff8000001a40c0: 0xffff8000001a41f0 0xffff800000180768
bootstack _ZN4core10intrinsics23is_aligned_and_not_null17hb8cd0a2694775395E
ffff8000001a40d0: 0xffff8000001a41f0 0xffff800000198fae
ffff8000001a40e0: 0x0000000000000001 0xffff800000198fae
ffff8000001a40f0: 0x0000000000000001 0xffff800000198fae
ffff8000001a4100: 0x0000000000000001 0x0000000000000001
ffff8000001a4110: 0x0000000000000001 0xffff800000171aaf
ffff8000001a4120: 0xffff800000198fae 0xffff800000180628
ffff8000001a4130: 0xffff800000198fae 0x0100000000000000
ffff8000001a4140: 0xffff800000198fae 0xffff8000001a41f0
```
I'm sorta expecting these to all be zeroes, or at best maybe some control structure at the base of bootstack (I didn't see this anywhere however).

Running things again however with `cargo xtask qemukvm --release` gets me all the way until the intended panic.  Not sure the messages are outputting the way we expect them to here, but at least we know what my issue is now...

Sadly, if I replace the intended panic!() at the end of theon with a `loop { relax() }` instead of halting (so I can inspect memory), I see **almost** the same stack crash on the --release build:

```
(qemu) x/512xg 0xffff800000126000
ffff800000126000: 0x0000000000000000 0x0000000000000000
ffff800000126010: 0x0000000000000000 0x0000000000000000
ffff800000126020: 0x0000000000000000 0x0000000000000000
ffff800000126030: 0x0000000000000000 0x0000000000000000
ffff800000126040: 0x0000000000000000 0x0000000000000000
ffff800000126050: 0x0000000000000000 0x0000000000000000
ffff800000126060: 0x0068000000000000 0xffff800000120000
ffff800000126070: 0x0000000000000000 0xffff800000120000
ffff800000126080: 0x0000000000000001 0xffff800000120001
ffff800000126090: 0x0000000000000013 0xffff800000102ee0
ffff8000001260a0: 0x0000000000000000 0x0000000000000001
ffff8000001260b0: 0xffff800000118870 0xffff80000012003f
ffff8000001260c0: 0x000000000000001b 0x000000050000000f
ffff8000001260d0: 0xffff800000103040 0x0000000000000008
ffff8000001260e0: 0x0000000000000008 0xffff800000102ca0
ffff8000001260f0: 0xffff8000001200ba 0x0000000000000020
ffff800000126100: 0xffff8000001200fa 0x0000000000000014
ffff800000126110: 0x0000000500000014 0xffff800000120110
ffff800000126120: 0x0000000000000002 0xffff800000120112
```

this also looks like garbage codegen from the string formatting?
```
(gdb) disassemble 0xffff800000102ee0
Dump of assembler code for function _ZN42_$LT$$RF$T$u20$as$u20$core..fmt..Debug$GT$3fmt17h38ffe86ed05baea5E:
   0xffff800000102ee0 <+0>:     ret
(gdb) disassemble 0xffff800000118870
Dump of assembler code for function _ZN4core3fmt17pointer_fmt_inner17h47a9a4d82d2d3b13E:
```

I wonder what ways exist to shrink the stack usage for the formatting/json serialization?  Feels like a potentially unbounded problem without knowing more.

waych commented:
sigh now I see the stack is 8192.  Let me play a little more with this.

waych commented:
Hmm.  Yeah so it is definitely blowing the stack and stomping all over other things in the bss in the debug build.  With the release build I see the stack get dirty several hundred bytes into the second 4k page.

While playing with this, I thought it was interesting that the BSP hlting early via the panic really messed up the synchronization of the APs re that poor mans spinlock.  Compare the following which is my release build terminating with a BSP panic:

```
    ProgramHeader {                                                                                                                                                               p_type: "PT_LOAD",                                                                                                                                                        p_flags: 0x5,                                                                                                                                                             p_offset: 0x1000,                                                                                                                                                         p_vaddr: 0xfffff00000000000,                                                                                                                                              p_paddr: 0xfffff00000000000,                                                                                                                                              p_filesz: 0x8,                                                                                                                                                            p_memsz: 0x8,
        p_align: 0x1000,
    },
    ProgramHeader {
        p_type: "PT_GNU_STACK",
        p_flags: 0x6,
        p_offset: 0x0,
        p_vaddr: 0x0,
        p_paddr: 0x0,
        p_filesz: 0x0,
        p_memsz: 0x0,
        p_align: 0x0,
    },
]
starting APs
le
3PoA NfIrCo:m
*
*H*e*l*l*o* *f*ro*m* *2
]H*e [l lCou tf rHoemr e1
 *************
PanicInfo {
    payload: Any { .. },
    message: Some(
        main: trapstubs = 0xffff8000001137d0,
    ),
    location: Location {
        file: "theon/src/main.rs",
        line: 177,
        col: 5,                                                                                                                                                               },                                                                                                                                                                        can_unwind: true,                                                                                                                                                     }
******************************************                                                                                                                                System halted.
```

vs if I put a spin loop on an atomic to create a loop the compiler won't optimize away doing this instead of the panic!():
```
    // XXX: This will never return.
    static S: AtomicBool = AtomicBool::new(false);
    while S.compare_exchange(true, false, Ordering::SeqCst, Ordering::Acquire).is_err() {
        arch::cpu::relax();
    }
```

I get ordered output from the APs:
```
        p_memsz: 0x0,
        p_align: 0x0,
    },
]
starting APs
Hello from 3
Hello from 2
Hello from 1
```

dancrossnyc commented:
> fwiw I think the stack got smashed, sadly due to using the default debug build.

Sigh. The joys of kernel programming.... (The sound you hear is me over here cursing about the stack usage of the fmt stuff).

> Running info registers repeatedly while hung I see the RIP fluctuate throughout `trap()`, but **also** I see it sometimes as 0000000000000008 which is clearly not correct. Not sure where this is coming from but it explains what the cpu is doing. Looking at the stack however, it looks like it isn't smashed, the RIP looks like its in a good spot relative to the bootstack page, however:

This makes some sense to me: the stacks are adjacent in the bump allocator, so if the formatting stuff is descending down into another 4K page, it's smashing the stack of the CPU before it. I can totally see that causing a core to jump off into la-la land.

>[snip]
> 
> I'm sorta expecting these to all be zeroes, or at best maybe some control structure at the base of bootstack (I didn't see this anywhere however).
> 
> Running things again however with `cargo xtask qemukvm --release` gets me all the way until the intended panic. Not sure the messages are outputting the way we expect them to here, but at least we know what my issue is now...

Yeah, the BSP is almost deliberately racy with the APs.

> Sadly, if I replace the intended panic!() at the end of theon with a `loop { relax() }` instead of halting (so I can inspect memory), I see **almost** the same stack crash on the --release build:
> 
> ```
> (qemu) x/512xg 0xffff800000126000
> ffff800000126000: 0x0000000000000000 0x0000000000000000
> ffff800000126010: 0x0000000000000000 0x0000000000000000
> ffff800000126020: 0x0000000000000000 0x0000000000000000
> ffff800000126030: 0x0000000000000000 0x0000000000000000
> ffff800000126040: 0x0000000000000000 0x0000000000000000
> ffff800000126050: 0x0000000000000000 0x0000000000000000
> ffff800000126060: 0x0068000000000000 0xffff800000120000
> ffff800000126070: 0x0000000000000000 0xffff800000120000
> ffff800000126080: 0x0000000000000001 0xffff800000120001
> ffff800000126090: 0x0000000000000013 0xffff800000102ee0
> ffff8000001260a0: 0x0000000000000000 0x0000000000000001
> ffff8000001260b0: 0xffff800000118870 0xffff80000012003f
> ffff8000001260c0: 0x000000000000001b 0x000000050000000f
> ffff8000001260d0: 0xffff800000103040 0x0000000000000008
> ffff8000001260e0: 0x0000000000000008 0xffff800000102ca0
> ffff8000001260f0: 0xffff8000001200ba 0x0000000000000020
> ffff800000126100: 0xffff8000001200fa 0x0000000000000014
> ffff800000126110: 0x0000000500000014 0xffff800000120110
> ffff800000126120: 0x0000000000000002 0xffff800000120112
> ```
> 
> this also looks like garbage codegen from the string formatting?

Is that on the BSP? That's kinda weird.... My guess is that the BSP stack is actually adjacent in the BSS to the bump allocator (or, at least, not far from it: goblin will allocate some stuff, but it's probably small compared to the formatting machinery). It could well be that one of the APs is clobbering the BSP stack, causing the BSP to jump off into infinity.

> I wonder what ways exist to shrink the stack usage for the formatting/json serialization? Feels like a potentially unbounded problem without knowing more.
>
> Hmm. Yeah so it is definitely blowing the stack and stomping all over other things in the bss in the debug build. With the release build I see the stack get dirty several hundred bytes into the second 4k page.

This is something of a known issue. Let me ask some folks around here what they do about it for e.g. Hubris. In the meantime, I've expanded all the stacks to 32KiB, which "ought to be enough for now." (Famous last words...).
 
> While playing with this, I thought it was interesting that the BSP hlting early via the panic really messed up the synchronization of the APs re that poor mans spinlock. Compare the following which is my release build terminating with a BSP panic:
> 
>[snip]
> 
> vs if I put a spin loop on an atomic to create a loop the compiler won't optimize away doing this instead of the panic!():
>
> ```
>     // XXX: This will never return.
>     static S: AtomicBool = AtomicBool::new(false);
>     while S.compare_exchange(true, false, Ordering::SeqCst, Ordering::Acquire).is_err() {
>         arch::cpu::relax();
>     }
> ```
> 
> I get ordered output from the APs:
> 
> ```
>         p_memsz: 0x0,
>         p_align: 0x0,
>     },
> ]
> starting APs
> Hello from 3
> Hello from 2
> Hello from 1
> ```

Much nicer. :-)

I had kind of wanted to see the panic, mostly just to differentiate against some other failure, but there's no good way to make the BSP participate in the synchronization protocol the APs use to serialize their output (spinning on that atomic). Certainly, having the BSP spin forever will prevent it from racing with the APs during print, though.

Another interesting thing I just ran into locally in release mode is an AP escaping from its `hlt` jail and hitting the panic!() at the end of `apmain`. What I think is going on is that it's getting the second IPI from `startaps()` after it has halted, and that's kicking it out of the inactive state and causing it to continue execution. It strikes me that it is not entirely clear what the processor does when it receives a SIPI and is not in the "wait-for-SIPI" state; the SDM and AMD refs indicate that the AP should start executing the AP startup code (even a second time) but that's not what I've generally observed. Regardless, it seems best to follow the docs and not send a second SIPI if the AP is executing.

What all of this suggests to me is that the BSP should really be blocked waiting on the APs to finish initializing, and we should be more circumspect about how we send SIPIs. In particular, it would probably be best to have a counter that the BSP spins on, waiting for all APs to finish starting, and also share an array of booleans or something with the APs so that they can mark that they are awake very early in the boot sequence, so we avoid sending them the second IPI. I think the broadcast INIT and broadcast first-SIPI are fine, but then we should loop over all the APs, checking whether they've started booting, and only SIPI again if they have not. I'll make that change; it'll be more robust anyhow.

dancrossnyc commented:
Ok, I've made the changes mentioned above. PTAL

waych commented on theon/src/x86_64/pc/boot.S:
```diff
@@ -412,23 +433,69 @@ eapstart:
+1:
+	lea	EXApicDisabled(%rip), %rdi
+	hlt
+	jmp	1b
```
Maybe leave a note to explain this now weirdly placed jmp statement?

dancrossnyc replied:
I've introduced a label `apearlypanic` with an explanatory comment, and jmp there instead of the loop. I also tidied up a bit more of the assembler there. PTAL

waych reviewed (approved):
Looks good.  Comments help the reader a lot here.

dancrossnyc commented:
> [snip]
> > I wonder what ways exist to shrink the stack usage for the formatting/json serialization? Feels like a potentially unbounded problem without knowing more.
> > Hmm. Yeah so it is definitely blowing the stack and stomping all over other things in the bss in the debug build. With the release build I see the stack get dirty several hundred bytes into the second 4k page.
> 
> This is something of a known issue. Let me ask some folks around here what they do about it for e.g. Hubris. In the meantime, I've expanded all the stacks to 32KiB, which "ought to be enough for now." (Famous last words...).

The close the loop on this.... I checked in with Cliff Biffle about this. His comment surprised me a bit, but debug builds are known to dramatically increase both stack usage and code size. The gist of it is that Debug sort of assumes "infinite-sized computers." While a bit tongue-in-cheek, the sentiment is accurate: no optimization at all yields excessive stack utilization.

The recommendation is to set:

```
[profile.dev]
opt-level = 1
```

in `Cargo.toml`, which seems reasonable. We'll still get debuggable code, but it won't be excessively inefficient. I'll make that change and send it out in a separate PR.

dancrossnyc commented:
BTW: Just for kicks I pumped up the number of CPUs Hypatia sees to 288 locally, which was the maximum I could get qemu to support. They all seemed to come up reliably.

waych commented:
FYI I think this is PR is still open maybe because you are using the repo directly and refreshed your smp branch?  Not sure if this is your intended desire.

dancrossnyc commented:
> FYI I think this is PR is still open maybe because you are using the repo directly and refreshed your smp branch? Not sure if this is your intended desire.

Oh, I just wanted to make sure you were ok with my last round of updates before submitting. Sorry if that wasn't clear!
